providers/class-two-factor-provider.php

```diff
@@ -72,6 +72,43 @@ public function pre_process_authentication( $user ) {
 	 */
 	abstract public function validate_authentication( $user );
 
+	/**
+	 * Logs the failed authentication.
+	 *
+	 * @param WP_User      $user WP_User object of the user trying to login.
+	 * @param string|false $code The code used to authenticate, if available.
+	 *
+	 * @since 0.9.0
+	 *
+	 * @return void
+	 */
+	public function log_failure( $user, $code = false ) {
+		/**
+		 * This action is triggered when a Two Factor validation fails.
+		 *
+		 * @param WP_User             $user WP_User object of the user trying to login.
+		 * @param string|false        $code The code used to authenticate, if available.
+		 * @param Two_Factor_Provider $this The Provider class used during the authentication.
+		 */
+		do_action( 'two_factor_user_login_failed', $user, $code, $this );
+
+		/* translators: %1$d: the user's ID %2$s: the code used to authenticate */
+		$log_message = sprintf( esc_html__( 'The user with ID %1$d failed to login using the code "%2$s"', 'two-factor' ), $user->ID, esc_html( $code ) );
+
+		/**
+		 * This filter is triggered to checke whether it's needed to log the authentication failure.
+		 *
+		 * @param boolean      $should_log  Whether or not the authentication failure should be logged.
+		 * @param WP_User      $user        WP_User object of the user trying to login.
+		 * @param string|false $code        The code used to authenticate, if available.
+		 * @param string       $log_message The generated log message.
+		 * @param Two_Factor_Provider $this The Provider class used during the authentication.
+		 */
+		if ( apply_filters( 'two_factor_log_failure', true, $user, $code, $log_message, $this ) ) {
+			error_log( $log_message );
```

Comment:
@jeffpaul, @georgestephanis What are your thoughts on this?

Reply:
My only uncertainty is if it's a good idea to include the bad code in this.  I don't have any strong opposition, but I'm slightly concerned that if it's one-digit off, or enough times combined with the timestamp being one tick off, something may be able to be reversed.

But I'm fine with email or not (someone trying to brute force the two factor code may be ... uhh ... a lot of emails though) -- but I definitely think error_log should be included.

Also, appreciate the filter so a third-party catcher could also listen for failures for alternate logs.

Reply:
Maybe support the admin_email but have it disabled and filterable to enable?  I think I'd want it enabled on sites I'm using, but I could see where that could get VERY chatty for someone with a large amount of users?

Reply:
It might also depend on the provider; for TOTP it's probably not uncommon to make a typo, and getting an email for that seems like overkill. Emailing after 3 failed attempts could be more reasonable (as long as there's a cap on how often it's sent), but IMO #477 seems like a better way to address brute force attacks.

If a backup code fails, though, I can see it being more useful to email, because it's very out of the ordinary. Related #476. 

WebAuthn keys (#427) could also be worth emailing about, since they should be less prone to accidental failure, especially with user verification/PINs disabled.

-------------------------------

[I don't think gaining access to several TOTPs would let you reverse the secret](https://security.stackexchange.com/a/162367/8467), but I'm not sure how useful including it is either.

Reply:
I'm afraid that moving the logging function out of the providers will not allow many customizations from them. I can think of some examples where it's useful to allow providers to set their own log data:

1. An email provider can send the email address used to log in instead of the user ID.
2. A provider could define their own "failed amount" before logging in.